daft/runners/ray_metrics.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+import dataclasses
+import logging
+import threading
+from collections import defaultdict
+
+logger = logging.getLogger(__name__)
+
+try:
+    import ray
+except ImportError:
+    logger.error(
+        "Error when importing Ray. Please ensure that getdaft was installed with the Ray extras tag: getdaft[ray] (https://www.getdaft.io/projects/docs/en/latest/learn/install.html)"
+    )
+    raise
+
+METRICS_ACTOR_NAME = "metrics"
+METRICS_ACTOR_NAMESPACE = "daft"
+
+
+@dataclasses.dataclass(frozen=True)
+class TaskEvent:
+    task_id: str
+
+
+@dataclasses.dataclass(frozen=True)
+class StartTaskEvent(TaskEvent):
+    """Marks the start of a task, along with available metadata"""
+
+    # Start Unix timestamp
+    start: float
+
+    # What stage this task belongs to
+    stage_id: int
```

Comment:
this can also be a `PartitionTaskContext`

Reply:
This one is a little difficult to unify - I decided not to unify these concepts. The `PartitionTaskContext` is left as a runtime concept, but when logging events the data in the events don't really fit into it all the time